src/Uno.UI/Microsoft/UI/Xaml/Controls/TreeView/TreeViewViewModel.cs

```diff
@@ -0,0 +1,964 @@
+				}
+			}
+
+			return descendantCount;
+		}
+
+		private int IndexOfNextSibling(TreeViewNode childNode)
+		{
+			var child = childNode;
+			var parentNode = child.Parent;
+			int stopIndex;
+			bool isLastRelativeChild = true;
+			while (parentNode != null && isLastRelativeChild)
+			{
+				int relativeIndex = parentNode.Children.IndexOf(childNode);
+				if (parentNode.Children.Count - 1 != relativeIndex)
+				{
+					isLastRelativeChild = false;
+				}
+				else
+				{
+					child = parentNode;
+					parentNode = parentNode.Parent;
+				}
+			}
+
+			if (parentNode != null)
+			{
+				int siblingIndex = parentNode.Children.IndexOf(childNode);
+				var siblingNode = parentNode.Children[siblingIndex + 1];
+				IndexOfNode(siblingNode, out stopIndex);
+			}
+			else
+			{
+				stopIndex = Count;
+			}
+
+			return stopIndex;
+		}
+
+		private int GetExpandedDescendantCount(TreeViewNode parentNode)
+		{
+			var allOpenedDescendantsCount = 0;
+			for (var i = 0; i < parentNode.Children.Count; i++)
+			{
+				var childNode = parentNode.Children[i];
+				allOpenedDescendantsCount++;
+				if (childNode.IsExpanded)
+				{
+					allOpenedDescendantsCount += CountDescendants(childNode);
+				}
+			}
+			return allOpenedDescendantsCount;
+		}
+
+		internal bool IsNodeSelected(TreeViewNode targetNode)
+		{
+			return m_selectedNodes.IndexOf(targetNode) > -1;
+		}
+
+		private TreeNodeSelectionState NodeSelectionState(TreeViewNode targetNode)
+		{
+			return targetNode.SelectionState;
+		}
+
+		private void UpdateNodeSelection(TreeViewNode selectNode, TreeNodeSelectionState selectionState)
+		{
+			var node = selectNode;
+			if (selectionState != node.SelectionState)
+			{
+				node.SelectionState = selectionState;
+				var selectedNodes = m_selectedNodes;
+				switch (selectionState)
+				{
+					case TreeNodeSelectionState.Selected:
+						selectedNodes.InsertAtCore(selectedNodes.Count, selectNode);
+						selectNode.ChildrenChanged += SelectedNodeChildrenChanged;
+						break;
+
+					case TreeNodeSelectionState.PartialSelected:
+					case TreeNodeSelectionState.UnSelected:
+						int index = selectedNodes.IndexOf(selectNode);
+						if (index > -1)
+						{
+							selectedNodes.RemoveAtCore(index);
+							selectNode.ChildrenChanged -= SelectedNodeChildrenChanged;
+						}
+						break;
+				}
+			}
+		}
+
+		internal void UpdateSelection(TreeViewNode selectNode, TreeNodeSelectionState selectionState)
+		{
+			if (NodeSelectionState(selectNode) != selectionState)
+			{
+				UpdateNodeSelection(selectNode, selectionState);
+
+				if (!IsInSingleSelectionMode())
+				{
+					UpdateSelectionStateOfDescendants(selectNode, selectionState);
+					UpdateSelectionStateOfAncestors(selectNode);
+				}
+			}
+		}
+
+		private void UpdateSelectionStateOfDescendants(TreeViewNode targetNode, TreeNodeSelectionState selectionState)
+		{
+			if (selectionState == TreeNodeSelectionState.PartialSelected) return;
```

Comment:
Codacy found an issue: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5156819)